ray_provider/decorators/ray_decorators.py

```diff
@@ -31,42 +45,180 @@ def ray_task(
     eager: bool = False,
 ):
     """Wraps a function to be executed on the Ray cluster.
-
-    The return values of the function will be cached on the Ray object store.
-    Downstream tasks must be ray tasks too, as the dependencies will be
-    fetched from the object store. The RayBackend will need to be setup in your
-    Dockerfile to use this decorator.
-
-    Use as a task decorator: ::
-
-        from ray_provider.decorators import ray_task
-
-        def ray_example_dag():
-
-            @ray_task("ray_conn_id")
-            def sum_cols(df: pd.DataFrame) -> pd.DataFrame:
-                return pd.DataFrame(df.sum()).T
-
-    :param python_callable: Function to be invoked on the Ray cluster.
-    :type python_callable: Optional[Callable]
-    :param http_conn_id: Http connection id for conenction to ray.
-    :type http_conn_id: str
-    :param ray_worker_pool: The pool that controls the
-            amount of parallel clients created to access the Ray cluster.
-    :type ray_worker_pool: Optional[str]
-    :param eager: Whether to run the the function on the
-            coordinator process (on the Ray cluster) or to
-            send the function to a remote task. You should
-            set this to False normally.
-    :type eager: Optional[bool]
     """
 
+    @provide_session
+    def on_retry_callback(context, session=None):
+        """When a task is set to retry, store the output of its upstream tasks.
+        """
+
+        # List upstream task ids
+        upstream_tasks = RayPythonOperator._upstream_tasks(
+            context.get('ti').task_id, context.get('dag'))
+
+        # Retrieve upstream object ids from xcom
+        upstream_objects = [RayPythonOperator._retrieve_obj_id_from_xcom(
+            task_id, context.get('dag').dag_id) for task_id in upstream_tasks]
+
+        # Retrieve the KV Actor
+        actor_ray_kv_store = KVStore("ray_kv_store").get_actor("ray_kv_store")
+
+        # Write to GCS
+        for dag_id, task_id, obj_ref in upstream_objects:
+            actor_ray_kv_store.gcs_dump.remote(dag_id, task_id, obj_ref)
+
     @functools.wraps(python_callable)
     def wrapper(f):
 
         return task(
             ray_wrapped(f, ray_conn_id, eager=eager),
             pool=ray_worker_pool,
+            on_retry_callback=on_retry_callback
         )
 
     return wrapper
+
+
+class RayPythonOperator(PythonOperator):
+
+    def __init__(self, *,
+                 python_callable: Callable,
+                 op_args: Optional[List] = None,
+                 op_kwargs: Optional[Dict] = None,
+                 templates_dict: Optional[Dict] = None,
+                 templates_exts: Optional[List[str]] = None,
+                 **kwargs) -> None:
```

Comment:
should we separate this to another PR?

Reply:
Oh, this is basically the implementation of the task?

Reply:
Can you leave a docstring here?

Reply:
Previously, the ray decorator used the `PythonOperator`. 

Assigning the recovered objects as Task attributes requires modifying to the `pre_execute` method of the `PythonOperator`. 

`RayPythonOperator` subclasses the `PythonOperator` and sets a `pre_execute` method and implements [logic to enable attribute assignment](https://github.com/anyscale/airflow-provider-ray/blob/1a515b9b2936c2d7c48eba6361618abda179a540/ray_provider/decorators/ray_decorators.py#L92).

- [ ] I will add docstrings explaining the implementation.